wgsl/index.bs

```diff
@@ -931,9 +941,230 @@ a result value, such as the non-result-value effects of its subexpressions.
 
 TODO: example: non-result-value effect is any side effect of a function call subexpression.
 
+## Ideal Types ## {#ideal-types-sec}
+
+There are two <dfn noexport>ideal types</dfn>:
+* The <dfn noexport>IdealInteger</dfn> type is the set of all integers.
+    * An [=syntax/int_literal=] without an `i` or `u` suffix denotes an [=IdealInteger=] value.
+* The <dfn noexport>IdealReal</dfn> type is the set of all real numbers (e.g. 1.25, -100, &pi;)
+    * A [=syntax/float_literal=] without an `f` suffix denotes a [=IdealReal=] value.
+
+A type is <dfn noexport>concrete</dfn> if it is not [=ideal type|ideal=].
+
+A [=numeric literal=] without a suffix denotes a value in an ideal type.
+However, each numeric value in memory, in pipeline I/O, or used in most
+computations belongs to one of the [=numeric scalar=] types.
+
+Converting a value from an ideal type to a numeric scalar type is called <dfn noexport>materialization</dfn>.
+* An [=IdealReal=] value can only convert to [=f32=].
+* An [=IdealInteger=] value can only convert to one of the [=numeric scalar=] types.
+
+<dfn noexport>Materialization Rule</dfn>: An expression |E| of [=ideal type=] is
+converted to a [=numeric scalar=] type |T| to satisfy type rules. This occurs when:
+    * |E| is the right-hand-side of a let-declaration with explicit type |T|.
+    * |E| is the right-hand-side of a var declaration with explicit [=store type=] |T|.
```

Comment:
```suggestion
    * |E| is the right-hand-side of a var declaration with explicit [=store type=] |T|.
    * |E| is the right-hand-side of an [=statement/assignment=] statement to a variable with [=store type=]|T|.
```

The assignment statement will always have a well-typed LHS that can be used to infer the type of literals on the RHS.